File: solar_clock.py
# solar_clock.py
#
# A Python script for a conceptual solar clock.
#
# To run this script, you first need to install the 'astral' library:
# pip install astral
#
# Then, execute the script from your terminal:
# python solar_clock.py
#
# The clock will start running and updating in place. Press Ctrl+C to stop.

import time
from datetime import datetime, timezone
import sys

try:
    # astral is used to get precise dates for solstices and equinoxes
    from astral.sun import equinoxes_and_solstices
except ImportError:
    print("Error: The 'astral' library is required.")
    print("Please install it using: pip install astral")
    sys.exit(1)

# --- Define the custom time system based on the prompt ---
# These constants define the structure of the custom time system.
WICKS_PER_MINUTE = 9
MINUTES_PER_HOUR = 7
HOURS_PER_DAY = 8
DAYS_PER_WEEK = 5
WEEKS_PER_MONTH = 9
MONTHS_PER_AGE = 3

# --- Pre-calculate total "wicks" for each period for efficiency ---
# This avoids recalculating these values in every loop.
WICKS_PER_HOUR = WICKS_PER_MINUTE * MINUTES_PER_HOUR
WICKS_PER_DAY = WICKS_PER_HOUR * HOURS_PER_DAY
WICKS_PER_WEEK = WICKS_PER_DAY * DAYS_PER_WEEK
WICKS_PER_MONTH = WICKS_PER_WEEK * WEEKS_PER_MONTH
WICKS_PER_AGE = WICKS_PER_MONTH * MONTHS_PER_AGE


def get_celestial_events(year):
    """
    Gets the UTC datetimes for the equinoxes and solstices of a given year.
    The events are returned in chronological order.
    """
    # The astral library provides the four key moments of the solar year.
    events = equinoxes_and_solstices(year)
    return sorted(events, key=lambda x: x.timestamp())


def find_current_age(now, events_this_year, events_next_year):
    """
    Determines the current age (season), its number, start time, and end time.
    Handles the transition between years seamlessly.
    """
    all_events = events_this_year + [events_next_year[0]]

    # Case 1: Before the first event of the year (Vernal Equinox).
    # This means we are in the 'Winter' age that began in the previous year.
    if now < all_events[0]:
        events_last_year = get_celestial_events(now.year - 1)
        start_event = events_last_year[3]  # Previous year's Winter Solstice
        end_event = all_events[0]          # This year's Vernal Equinox
        age_name = "Winter"
        age_number = 4
    # Case 2: Between Vernal Equinox and Summer Solstice.
    elif all_events[0] <= now < all_events[1]:
        age_name = "Spring"
        age_number = 1
        start_event = all_events[0]
        end_event = all_events[1]
    # Case 3: Between Summer Solstice and Autumnal Equinox.
    elif all_events[1] <= now < all_events[2]:
        age_name = "Summer"
        age_number = 2
        start_event = all_events[1]
        end_event = all_events[2]
    # Case 4: Between Autumnal Equinox and Winter Solstice.
    elif all_events[2] <= now < all_events[3]:
        age_name = "Autumn"
        age_number = 3
        start_event = all_events[2]
        end_event = all_events[3]
    # Case 5: After the Winter Solstice.
    else:
        age_name = "Winter"
        age_number = 4
        start_event = all_events[3]
        end_event = all_events[4]  # The Vernal Equinox of the next year.

    return age_name, age_number, start_event, end_event


def main():
    """Main function to run the solar clock."""
    print("Initializing Solar Clock...")
    last_year_checked = -1
    events_this_year = None
    events_next_year = None

    try:
        while True:
            now = datetime.now(timezone.utc)

            # To save resources, only fetch the celestial events when the year changes.
            if now.year != last_year_checked:
                print(f"Calibrating for the year {now.year}...")
                last_year_checked = now.year
                events_this_year = get_celestial_events(now.year)
                events_next_year = get_celestial_events(now.year + 1)

            # Determine the current age and its start/end boundaries.
            age_name, age_num, age_start, age_end = find_current_age(now, events_this_year, events_next_year)

            # Calculate the total duration of the current age in real-world seconds.
            total_age_duration_sec = (age_end - age_start).total_seconds()
            # Calculate how many real-world seconds have passed in this age.
            elapsed_sec_in_age = (now - age_start).total_seconds()

            # Find what fraction of the age has been completed.
            age_proportion = elapsed_sec_in_age / total_age_duration_sec

            # Convert the proportion of the age passed into the total number of "wicks" passed.
            total_wicks = int(age_proportion * WICKS_PER_AGE)

            # Deconstruct the total wicks into the custom time units using division and modulo.
            remaining_wicks = total_wicks
            
            # Note: Months, Weeks, and Days are 1-based (start from 1).
            months = (remaining_wicks // WICKS_PER_MONTH) + 1
            remaining_wicks %= WICKS_PER_MONTH
            
            weeks = (remaining_wicks // WICKS_PER_WEEK) + 1
            remaining_wicks %= WICKS_PER_WEEK
            
            days = (remaining_wicks // WICKS_PER_DAY) + 1
            remaining_wicks %= WICKS_PER_DAY
            
            # Note: Hours, Minutes, and Wicks are 0-based (start from 0).
            hours = remaining_wicks // WICKS_PER_HOUR
            remaining_wicks %= WICKS_PER_HOUR
            
            minutes = remaining_wicks // WICKS_PER_MINUTE
            wicks = remaining_wicks % WICKS_PER_MINUTE
            
            # --- Format and display the output ---
            # Format: AgeNumber.Month.Week.Day.Hour.Minute.Wick
            time_str = f"{age_num}.{months}.{weeks}.{days}.{hours}.{minutes}.{wicks}"
            
            # Use carriage return '\r' to print on the same line, creating an updating effect.
            # Use sys.stdout.flush() to ensure the output is shown immediately.
            display_string = f"\r{age_name} Age | Time: {time_str:<20}" # Padding for clean overwriting
            sys.stdout.write(display_string)
            sys.stdout.flush()

            # Sleep for a fraction of a second to create a running clock effect.
            time.sleep(1/9)

    except KeyboardInterrupt:
        print("\nSolar clock stopped.")
    except Exception as e:
        print(f"\nAn error occurred: {e}")


if __name__ == "__main__":
    main()

File: solar_clock.c
/*
 * solar_clock.c
 *
 * A C implementation of the conceptual solar clock.
 *
 * This program requires a C compiler (like GCC) and the standard C library.
 *
 * How to Compile and Run:
 * 1. Save this file as solar_clock.c
 * 2. Open your terminal and compile the program using GCC:
 * gcc solar_clock.c -o solar_clock -lm
 * (The '-lm' flag is important to link the math library for floor())
 *
 * 3. Run the compiled program:
 * ./solar_clock
 *
 * The clock will start running. Press Ctrl+C to stop.
 */

#include <stdio.h>
#include <time.h>
#include <unistd.h> // For usleep()
#include <math.h>   // For floor()
#include <signal.h> // For signal handling (Ctrl+C)
#include <stdlib.h> // For exit()

// --- Define the custom time system ---
const int WICKS_PER_MINUTE = 9;
const int MINUTES_PER_HOUR = 7;
const int HOURS_PER_DAY = 8;
const int DAYS_PER_WEEK = 5;
const int WEEKS_PER_MONTH = 9;
const int MONTHS_PER_AGE = 3;

// --- Pre-calculate total "wicks" for each period ---
const long WICKS_PER_HOUR = WICKS_PER_MINUTE * MINUTES_PER_HOUR;
const long WICKS_PER_DAY = WICKS_PER_HOUR * HOURS_PER_DAY;
const long WICKS_PER_WEEK = WICKS_PER_DAY * DAYS_PER_WEEK;
const long WICKS_PER_MONTH = WICKS_PER_WEEK * WEEKS_PER_MONTH;
const long WICKS_PER_AGE = WICKS_PER_MONTH * MONTHS_PER_AGE;


// Converts a date string (like "2025-03-20T09:01:00Z") to a time_t timestamp.
time_t get_utc_timestamp(const char* date_string) {
    struct tm tm = {0};
    // strptime parses the string into the tm struct
    strptime(date_string, "%Y-%m-%dT%H:%M:%SZ", &tm);
    // timegm converts the tm struct to a UTC timestamp
    return timegm(&tm);
}

// Function to handle Ctrl+C for a clean exit
void handle_sigint(int sig) {
    printf("\nSolar clock stopped.\n");
    exit(0);
}

int main() {
    // Register the signal handler for Ctrl+C
    signal(SIGINT, handle_sigint);

    // --- Celestial Event Timestamps for 2025/2026 (in UTC) ---
    // These are the key moments that define the "Ages".
    const time_t DEC_SOLSTICE_2024_TS = get_utc_timestamp("2024-12-21T09:20:00Z");
    const time_t MAR_EQUINOX_2025_TS = get_utc_timestamp("2025-03-20T09:01:00Z");
    const time_t JUN_SOLSTICE_2025_TS = get_utc_timestamp("2025-06-21T02:42:00Z");
    const time_t SEP_EQUINOX_2025_TS = get_utc_timestamp("2025-09-22T18:19:00Z");
    const time_t DEC_SOLSTICE_2025_TS = get_utc_timestamp("2025-12-21T14:03:00Z");
    const time_t MAR_EQUINOX_2026_TS = get_utc_timestamp("2026-03-20T14:45:00Z");

    printf("Initializing Solar Clock for 2025...\n");
    printf("Press Ctrl+C to stop.\n");

    // Main loop
    while (1) {
        time_t now_ts;
        time(&now_ts); // Get current time as a timestamp

        // Variables to hold the current age's properties
        const char* age_name;
        int age_num;
        time_t age_start_ts;
        time_t age_end_ts;

        // --- Determine the current age ---
        if (now_ts < MAR_EQUINOX_2025_TS) {
            age_name = "Winter";
            age_num = 4;
            age_start_ts = DEC_SOLSTICE_2024_TS;
            age_end_ts = MAR_EQUINOX_2025_TS;
        } else if (now_ts >= MAR_EQUINOX_2025_TS && now_ts < JUN_SOLSTICE_2025_TS) {
            age_name = "Spring";
            age_num = 1;
            age_start_ts = MAR_EQUINOX_2025_TS;
            age_end_ts = JUN_SOLSTICE_2025_TS;
        } else if (now_ts >= JUN_SOLSTICE_2025_TS && now_ts < SEP_EQUINOX_2025_TS) {
            age_name = "Summer";
            age_num = 2;
            age_start_ts = JUN_SOLSTICE_2025_TS;
            age_end_ts = SEP_EQUINOX_2025_TS;
        } else if (now_ts >= SEP_EQUINOX_2025_TS && now_ts < DEC_SOLSTICE_2025_TS) {
            age_name = "Autumn";
            age_num = 3;
            age_start_ts = SEP_EQUINOX_2025_TS;
            age_end_ts = DEC_SOLSTICE_2025_TS;
        } else {
            age_name = "Winter";
            age_num = 4;
            age_start_ts = DEC_SOLSTICE_2025_TS;
            age_end_ts = MAR_EQUINOX_2026_TS;
        }

        // --- Calculate the custom time ---
        double total_age_duration_sec = difftime(age_end_ts, age_start_ts);
        double elapsed_sec_in_age = difftime(now_ts, age_start_ts);
        
        double age_proportion = elapsed_sec_in_age / total_age_duration_sec;
        
        long total_wicks = (long)floor(age_proportion * WICKS_PER_AGE);
        
        long remaining_wicks = total_wicks;

        // Deconstruct total wicks into our custom time units
        int months = (remaining_wicks / WICKS_PER_MONTH) + 1;
        remaining_wicks %= WICKS_PER_MONTH;

        int weeks = (remaining_wicks / WICKS_PER_WEEK) + 1;
        remaining_wicks %= WICKS_PER_WEEK;

        int days = (remaining_wicks / WICKS_PER_DAY) + 1;
        remaining_wicks %= WICKS_PER_DAY;

        int hours = remaining_wicks / WICKS_PER_HOUR;
        remaining_wicks %= WICKS_PER_HOUR;

        int minutes = remaining_wicks / WICKS_PER_MINUTE;
        int wicks = remaining_wicks % WICKS_PER_MINUTE;

        // --- Format and display the output ---
        // Use carriage return '\r' to print on the same line.
        // The padding with spaces at the end clears any leftover characters.
        printf("\r%-10s Age | Time: %d.%d.%d.%d.%d.%d.%d      ", 
               age_name, age_num, months, weeks, days, hours, minutes, wicks);
        
        // Flush the output buffer to ensure the line is printed immediately.
        fflush(stdout);
        
        // Sleep for 100,000 microseconds (0.1 seconds)
        usleep(100000);
    }

    return 0;
}

